python/pylibcudf/pylibcudf/strings/combine.pyx

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+from libcpp.memory cimport unique_ptr
+from libcpp.utility cimport move
+from pylibcudf.column cimport Column
+from pylibcudf.libcudf.column.column cimport column
+from pylibcudf.libcudf.scalar.scalar cimport string_scalar
+from pylibcudf.libcudf.strings cimport combine as cpp_combine
+from pylibcudf.scalar cimport Scalar
+from pylibcudf.table cimport Table
+
+from cython.operator import dereference
+from pylibcudf.libcudf.strings.combine import \
+    output_if_empty_list as OutputIfEmptyList  # no-cython-lint
+from pylibcudf.libcudf.strings.combine import \
+    separator_on_nulls as SeparatorOnNulls  # no-cython-lint
+
+
+cpdef Column concatenate(
+    Table strings_columns,
+    ColumnOrScalar separator,
+    Scalar narep,
+    Scalar col_narep,
+    separator_on_nulls separate_nulls,
+):
+    """
+    Concatenates all strings in the column into one new string delimited
+    by an optional separator string.
+
+    Parameters
+    ----------
+    strings_columns : Table
+        Strings for this operation
+
+    separator : Column or Scalar
+        Separator(s) for a given row
+
+    narep : Scalar
+        String to replace a null separator for a given row.
+
+    col_narep : Scalar
+        String that should be used in place of any null strings found in any column.
+        Ignored when separator is a Scalar.
```

Comment:
This is very confusing docstring. I do wonder if this is one of the cases where we can't do a good job of matching the overloaded C++ function.

In the case that `separator` is a `Scalar`, my reading of this docstring is that `narep` is used in when the separator is null. and I can't control how to replace null strings.
However, reading the code, in this case `narep` is used to replace null strings.

Reply:
I think @brandon-b-miller and I encountered something similar in a few other cases. I don't recall exactly which ones, but perhaps he does. We came up with a few different solutions to make Pythonic APIs that were close enough in spirit to the libcudf APIs. In some cases we bundled parameters into a tuple, or a similar structure that allowed grouping parameters that had to be provided together. Another option would be to use default `None` parameters with more informative errors when incompatible arguments are provided.

Reply:
How it's written here seems pretty close however I think it should error rather than ignore if an invalid combination of arguments is passed. 

For column factories we considered providing the mask and null count parameters in a `tuple`, but this was partly motivated by the desire to make an similar change in libcudf where apis would similarly be adjusted to expect a `pair`, I'm not sure there's quite the same symmetry here.

Reply:
Ah right, that's what it was. Yeah that doesn't offer much guidance here then. I agree with erroring rather than ignoring, for sure.